loading.bs

```diff
@@ -331,6 +371,304 @@ Note: Applying the signed exchange's response here has the effect of letting a
+
+    9. Let |allowed alternate subresource signed exchange links| be null.
+
+    10. If |sourceBrowsingContext|'s [=active document=]'s
+        [=Document/prefetched signed exchanges for navigation=] has a matching
+        exchange for the |request|'s [=request/url=], then:
+
+        1. Let |clonedExchange| be the result of creating a new [=exchange=]
+            with the matching exchange's [=exchange/request URL=] and the result
+            of [=response/cloning=] the matching exchange's
+            [=exchange/response=].
+
+        1. Move |clonedExchange| to the |request|'s
+            [=request/stashed exchange=].
+
+        1. Copy <{link/rel/allowed-alt-sxg}> link of the exchange's
+            [=exchange/response=] to
+            |allowed alternate subresource signed exchange links|.
+
+
+
+And before
+
+> 17. Run process a navigate response with navigationType, the source browsing
+>     context, and navigationParams.
+
+add the following steps:
+
+    19. Copy |sourceBrowsingContext|'s [=active document=]'s
+        [=Document/prefetched subresource signed exchanges=]
+        to |navigationParams|'s
+        [=navigation params/prefetched subresource signed exchanges=].
+
+    20. Set |navigationParams|'s
+        [=navigation params/allowed alternate subresource signed exchange links=]
+        to |allowed alternate subresource signed exchange links|.
+
+Note: As browsers move toward partitioned HTTP caches, the source document's
+    cache will likely be separate from the target's cache, so we can't just pass
+    prefetched content through the cache.
+
+
+<h3 algorithm id="mp-initialise-the-document-object">
+Monkeypatch Initialize a Document object</h3>
+In
+[initialize a Document object](https://html.spec.whatwg.org/multipage/browsing-the-web.html#initialise-the-document-object)
+before
+
+> 15. Return document.
+
+add the following steps:
+
+    15. Move the |navigationParams|'s
+        [=navigation params/prefetched subresource signed exchanges=] to the
+        |document|'s [=Document/transferred subresource signed exchanges=].
+
+    16. Set |document|'s
+        [=Document/allowed alternate subresource signed exchange links=] to
+        |navigationParams|'s
+        [=navigation params/allowed alternate subresource signed exchange links=].
+
+
+<h3 algorithm id="mp-link-type-preload">
+Monkeypatch Link type "preload"</h3>
+
+The navigated-to document has a set of preloads for which it uses the
+<{link/rel/allowed-alt-sxg}> link relation to declare that they can be served by
+signed exchanges. The UA either serves all of them from SXGs prefetched by the
+previous page, or none of them. So in the
+[linked resource fetch setup steps for preload type of linked resource](https://html.spec.whatwg.org/multipage/links.html#link-type-preload:linked-resource-fetch-setup-steps)
+before
+
+> 5. Return true.
+
+add the following steps:
+
+    5. If |el| is for the main resource's HTTP <a http-header>Link</a> header,
```

Comment:
I've run out of time to review today, so this comment is just a general direction rather than actually what needs to be fixed, but:

We're going to have to reconcile that the HTML spec pretends that it's dealing with `<link>` elements, but we're doing `Link` headers, and that HTML is trying to handle things incrementally, but we're putting a restriction on the overall set of `Link` headers. It might work to define a predicate on the response that runs over the full set of response headers to say whether to do the resource substitution at all, and then maybe you can avoid checking whether each element came from a header...

Reply:
Added note about the Issue(whatwg/html#4224).

Reply:
Also "is for the main resource" is not really precise, and it's be better to have some boolean that indicates what you mean here.